### ISSUE TITLE
Make populse_db as general as possible

### ISSUE THREAD

**Issue by ouvrierl** [populse/populse_db#18]
Does everyone agree for making the database as general as possible?

That would imply a table column (instead of tag), with only 2 columns: name (varchar), and type (varchar), as we need the values to be typed.

The table path would stay the same, it would just be renamed row.

The table initial would stay exactly the same.

Are you okay with not having the tag informations (origin, unit, default value, description) in the database? It would make them MIA2 specific (like the visibility).

We were also thinking about making a GUI for the database if it becomes general, but it's definitely not a priority.

**Comment by servoz** [comment]
ok for me !

**Comment by sapetnioc** [comment]
I totally agree on the main ideas. However, I have some comments we can
discuss :

I do not really like the name "row" for a table. What about "document" or
something like that ? The goal of populse_db would be to
store/retrieve/search documents.

We should keep a mandatory primary key column. Today it is the column
"name", shall we include it in the tag list ?

I would keep the possibility to add a description to a column. It can be
useful for user in graphical interface (for instance in a tooltip).

A bit off-topic : should we add the possibility to add index to column in
their definition ? Indexes can drasticaly improve query performances so it
can be useful for columns highly used in selections. We could add new
"index" optional parameter to add_column() (and the corresponding column in
"column" table). This may be done later and I hesitated to ask this
question in a separate issue.

Yann Cointepas            Tel: +33 1 69 08 78 31
CEA - Neurospin
Bâtiment 145, Point Courrier 156
91191 Gif-sur-Yvette cedex, France

**Comment by ouvrierl** [comment]
- I don't really mind the table name, document is fine with me.

- For the GUI, it would be easier to have the column "name" in the list of tags (columns). I currently have some problems in the MIA DataBrowser because of that (I can do it this afternoon).

- Yes for the tooltip we could keep the description it's a good idea

- Indexes create table views right? It would allow to take only the columns needed in path table? Would we really gain some performance? Since most of the time we query on the whole table.

**Comment by sapetnioc** [comment]
No, indexes does not change anything to the structure and usage of the
database. It justs add an internal structure (such as a b-tree) that makes
it possible to speed-up queries using the column (it can be a lot faster on
some cases). On the other hand, insertions can be slower with indexes
(especially for huge number of rows) and it consumes disk space and memory,
so we must not put indexes on all columns. There are various kind of
indexes. This is completely transparent for user. Columns with primaty keys
have an implicit index. On other columns it must be set explicitely with
SQL (for instance using CREATE INDEX).

**Comment by sapetnioc** [comment]
I had an idea I'd like to have your opinion on. If the database becomes generic, we will have the possibility to store/query all documents in a single place (I do not count the initial value table). What about adding the possibility to have several collections of documents ? Each collection would have its own set of column definition and its own set of documents (technically, a collection would be a table). 

The main difference in the API would be a few methods to manage collections such as:
```
new_collection(name)
collections()
delete_collection(name)
```
And existing methods would have a new `collection` parameter. For instance :
```
add_tag(collection, tag_name, tag_type, ...)
new_value(collection, ...)
```

To have the same features as today, we would create a single "path" collection and put all documents in it. But when we are ready to store processing history, we could add a new "processing_history" collection with possibly different columns.

I took the names "collection" and "document" from MongoDB (see the[ french Wikipedia page](https://fr.wikipedia.org/wiki/MongoDB). Surprisingly, english page has less information). If we want to be coherent with this naming, we could use "field" instead of "row".

**Comment by sapetnioc** [comment]
I think it could be interesting to have a meeting to discuss about a 1.0 version of the API. I think we have all implementation elements to review together the details API. I have a few questions/comments that I think may be easier to discuss without using a keyboard.

**Comment by ouvrierl** [comment]
I think the possibility to have collections is a great idea, but I don't know how we could use it efficiently in MIA, since we have strictly a database per project.

And I totally agree that it would be easier to do a meeting, as writing everything can take some time.

I have another question: the general populse_db is ready in my fork (I already pushed the beginning this morning on populse_db, I hope it didn't bother anyone). Should I put it on populse_db right now? On master or on a new branch? Because it changes a lot of code (even filter.py), as I renamed everything to have columns and documents. Since the functions are renamed, I also have to push populse_sandbox at the same time.

**Comment by ouvrierl** [comment]
For us it would be the best to do the meeting monday or tuesday morning, or tuesday after 4pm.
When are you available?

**Comment by sapetnioc** [comment]
Ok we will discuss all this. I am available on your three proposals.

I finished filter implementation on master. I think you can merge your changes on master. I have no open file and I will not do any change until your changes are merged. If you are not sure, you can do a pull request.

**Comment by ouvrierl** [comment]
Ok, I just merged everything.
When should we do the meeting?

**Comment by sapetnioc** [comment]
What about today ? Either this morning or at 2 p.m.

**Comment by ouvrierl** [comment]
We can say 10 a.m, in about 15 minutes if it's ok for you

**Comment by sapetnioc** [comment]
I am ready, what about you Denis ?

**Comment by denisri** [comment]
I'm here but don't find the connection address...

2018-05-28 9:56 GMT+02:00 Yann Cointepas <notifications@github.com>:

**Comment by sapetnioc** [comment]
What about closing this issue.
